test/IRGen/conditional-dead-strip-objc-ir.swift

```diff
@@ -6,18 +6,25 @@
 
 // RUN: %target-build-swift -Xfrontend -conditional-runtime-records %s -emit-ir -o - | %FileCheck %s
 
+import Foundation
+
 public class Class {
 }
 
+@objc public protocol ObjCAnnotatedProtocol {
+}
+
 // CHECK:      @llvm.{{(compiler.)?}}used = appending global [
 // CHECK-SAME:   @"$s4main5ClassCHn"
 // CHECK-SAME: ], section "llvm.metadata"
 
-// CHECK:      !llvm.used.conditional = !{[[M1:!.*]], [[M2:!.*]], [[O1:!.*]]}
+// CHECK:      !llvm.used.conditional = !{[[M1:!.*]], [[M2:!.*]], [[M3:!.*]], [[O1:!.*]]}
 
 // CHECK-DAG:      [[M1]]  = !{{{.*}} @"$s4main5ClassCMF", i32 0, [[M1A:!.*]]}
 // CHECK-DAG:      [[M1A]] = !{{{.*}} @"$s4main5ClassCMn"
-// CHECK-DAG:      [[M2]]  = !{{{.*}} @"$s4main5ClassCHn", i32 0, [[M1A:!.*]]}
+// CHECK-DAG:      [[M2]]  = !{{{.*}} @"$s4main21ObjCAnnotatedProtocol_pMF", i32 0, [[M2A:!.*]]}
+// CHECK-DAG:      [[M2A]] = !{{{.*}} @"$s4main21ObjCAnnotatedProtocolMp"
+// CHECK-DAG:      [[M3]]  = !{{{.*}} @"$s4main5ClassCHn", i32 0, [[M1A:!.*]]}
 
```

Comment:
Not important for the review, but for my own understanding I'd like to know what these metadata entities store. Swift demangle gives me:

```
~/git/swift-project swift-demangle s4main5ClassCMF
$s4main5ClassCMF ---> reflection metadata field descriptor main.Class
~/git/swift-project swift-demangle s4main5ClassCMn
$s4main5ClassCMn ---> nominal type descriptor for main.Class
~/git/swift-project swift-demangle s4main5ClassCHn
$s4main5ClassCHn ---> nominal type descriptor runtime record for main.Class
```
I assume a field descriptor lets you inspect the fields of an object at runtime? What is the difference between the type descriptor and type descriptor runtime record?

Reply:
See e.g. `emitTypeMetadataRecords` in GenDecl.cpp, which emits a list of these "runtime records" each of which is just a small record holding a reference (pointer / relative pointer) to the actual type descriptor, which is a larger data structure. The "runtime records" are placed into a special section (with a specific name) because at process startup that section is scanned/parsed by the runtime.

I don't know if it would be feasible to avoid having this indirection are literally place the type descriptors straight into the special section, but I imagine it's very desirable to have a small, fixed-sized, stable format of these "runtime records" to make the runtime processing of them simple and stable.